# exercise 4.5.

In [3]:
# importing libraries
import pandas as pd
import numpy as np
import os

In [2]:
# importing csv files
path = r'C:\Users\dog\desktop\Instacart Basket Analysis'

In [4]:
# importing csv file products
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [5]:
# importing csv file orderswrangled
df_orders_wrangled = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

In [6]:
# create a dataframe 
df_test = pd.DataFrame()

In [7]:
df_test['mix'] = ['a', 'b', 1, True]

In [8]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [9]:
# Check for mixed types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [10]:
df_test['mix'] = df_test['mix'].astype('str')

In [11]:
# Check for mixed types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

In [12]:
# Finding Missing Values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [13]:
# creating a subset data frame with  no values from column 
df_NaN = df_prods[df_prods['product_name'].isnull() == True]

In [14]:
df_NaN

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [15]:
df_prods.shape

(49693, 5)

In [16]:
# creating new data frame without NaN
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [17]:
df_prods_clean.shape

(49677, 5)

In [18]:
# Finding Duplicates
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [19]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [20]:
# Addressing Duplicates
df_prods_clean.shape

(49677, 5)

In [21]:
# dropping duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [22]:
df_prods_clean_no_dups.shape


(49672, 5)

In [23]:
# export checked prods data frame to csv
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'df_prods_clean_no_dups.csv'))

# Task 4.5.

In [24]:
# importing orders wrangled
df_orders_wrangled = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)


In [25]:
df_orders_wrangled.head()

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0           0   2539329        1             1                   2   
1           1   2398795        1             2                   3   
2           2    473747        1             3                   3   
3           3   2254736        1             4                   4   
4           4    431534        1             5                   4   

   order_hour_of_day  days_since_last_order  
0                  8                    NaN  
1                  7                   15.0  
2                 12                   21.0  
3                  7                   29.0  
4                 15                   28.0

In [26]:
# 2. checking for data frame prods unusual numbers. 

df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

#### There is max price of 99999, which seems to be a false value, however before checking up with client it is impossible to tell whether it is true or false

In [27]:
# 3. Check for mixed-type data in your df_ords dataframe.

df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'orders.csv'), index_col = False)

In [28]:
df_ords.head()

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2398795        1    prior             2          3                  7   
2    473747        1    prior             3          3                 12   
3   2254736        1    prior             4          4                  7   
4    431534        1    prior             5          4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [29]:
#3. Check for mixed-type data in your df_ords dataframe

for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

####  no mixed data type values were detected 

In [30]:
# 5. Run a check for missing values in your df_ords dataframe.
df_ords.isnull().sum()

order_id                       0
user_id                        0
eval_set                       0
order_number                   0
order_dow                      0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

In [31]:
df_ords.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3421083 entries, 0 to 3421082
Data columns (total 7 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   order_id                int64  
 1   user_id                 int64  
 2   eval_set                object 
 3   order_number            int64  
 4   order_dow               int64  
 5   order_hour_of_day       int64  
 6   days_since_prior_order  float64
dtypes: float64(1), int64(5), object(1)
memory usage: 182.7+ MB


In [32]:
df_ords.describe()

order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

In [33]:
# creating a subset of data with only NaN in order to check that
df_ords_days_since_prior_order_NaN = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [34]:
df_ords_days_since_prior_order_NaN.head(60)

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0     2539329        1    prior             1          2                  8   
11    2168274        2    prior             1          2                 11   
26    1374495        3    prior             1          1                 14   
39    3343014        4    prior             1          6                 11   
45    2717275        5    prior             1          3                 12   
50    2086598        6    prior             1          5                 18   
54    2565571        7    prior             1          3                  9   
75     600894        8    prior             1          6                  0   
79     280530        9    prior             1          1                 17   
83    1224907       10    prior             1          2                 14   
89    2411567       11    prior             1          0                 11   
97    1483446       12    prior             1          1                 12   
103   2618231       13    prior             1          6                 12   
116   2906263       14    prior             1          2                 18   
130   2514324       15    prior             1          5                 13   
153   1273676       16    prior             1          2                  8   
160   1737705       17    prior             1          2                 13   
201   2780889       18    prior             1          6                 18   
208   1321974       19    prior             1          6                 13   
218    947722       20    prior             1          1                 10   
223   2625565       21    prior             1          1                 10   
257   2300537       22    prior             1          3                  7   
273   1792521       23    prior             1          5                 14   
278    611214       24    prior             1          3                 20   
297   2610576       25    prior             1          1                 14   
301   3229714       26    prior             1          2                 15   
314    361493       27    prior             1          3                  9   
396   1138968       28    prior             1          0                 14   
421   2064090       29    prior             1          5                 11   
440    375303       30    prior             1          1                 13   
449   2150649       31    prior             1          5                 11   
470    134879       32    prior             1          3                 12   
476   2906103       33    prior             1          3                 14   
480   2383661       34    prior             1          4                 15   
486   1269089       35    prior             1          5                 21   
496   3285797       36    prior             1          4                 17   
534    327827       37    prior             1          2                 10   
558   2873174       38    prior             1          0                 10   
571   2782479       39    prior             1          0                 12   
579   2570028       40    prior             1          0                  9   
589    252244       41    prior             1          6                 16   
595    940313       42    prior             1          2                 10   
612    606677       43    prior             1          1                 18   
624   1359969       44    prior             1          0                  7   
628    300892       45    prior             1          1                 20   
633   3345812       46    prior             1          1                 16   
653   2857489       47    prior             1          5                 16   
659   1035233       48    prior             1          6                 16   
670   1216869       49    prior             1          2                  9   
679   1363380       50    prior             1          

#### as we can see, all the NaN have 1 amount of orders, which makes me conclude that it is the first order of that customer. In order to check that we need to find if there are any repeated user_id in the data frame with 1 order where the value is NaN AVG 11.11484

In [35]:
# 6 Address the missing values using an appropriate method
# mean_value = df_ords['days_since_prior_order'].mean()
# df_ords['days_since_prior_order'].fillna(mean_value, inplace=True)

df_ords_clean = df_ords
df_ords_clean['new_customer'] =df_ords['days_since_prior_order'].isnull() == True
df_ords_clean.head()


order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2398795        1    prior             2          3                  7   
2    473747        1    prior             3          3                 12   
3   2254736        1    prior             4          4                  7   
4    431534        1    prior             5          4                 15   

   days_since_prior_order  new_customer  
0                     NaN          True  
1                    15.0         False  
2                    21.0         False  
3                    29.0         False  
4                    28.0         False

#### the code above will replace the whole NaN values with average. This will not harm the whole description of the data. We can still identify the first customers by amount of orders in the order number column. In addition the code is always taking fresh data, since its not fixed, but always counted based on the csv file that will be imported. 

In [36]:
# checking if there is any NaN values left in the days_since_prior_order column
df_ords[df_ords['days_since_prior_order'].isnull() == True]

order_id  user_id eval_set  order_number  order_dow  \
0         2539329        1    prior             1          2   
11        2168274        2    prior             1          2   
26        1374495        3    prior             1          1   
39        3343014        4    prior             1          6   
45        2717275        5    prior             1          3   
...           ...      ...      ...           ...        ...   
3420930    969311   206205    prior             1          4   
3420934   3189322   206206    prior             1          3   
3421002   2166133   206207    prior             1          6   
3421019   2227043   206208    prior             1          1   
3421069   3154581   206209    prior             1          3   

         order_hour_of_day  days_since_prior_order  new_customer  
0                        8                     NaN          True  
11                      11                     NaN          True  
26                      14                     NaN          True  
39                      11                     NaN          True  
45                      12                     NaN          True  
...                    ...                     ...           ...  
3420930                 12                     NaN          True  
3420934                 18                     NaN          True  
3421002                 19                     NaN          True  
3421019                 15                     NaN          True  
3421069                 11                     NaN          True  

[206209 rows x 8 columns]

In [37]:
# 7 Run a check for duplicate values in your df_ords data.
df_ords_dups = df_ords[df_ords.duplicated()]

In [38]:
df_ords_dups.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 0 entries
Data columns (total 8 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   order_id                0 non-null      int64  
 1   user_id                 0 non-null      int64  
 2   eval_set                0 non-null      object 
 3   order_number            0 non-null      int64  
 4   order_dow               0 non-null      int64  
 5   order_hour_of_day       0 non-null      int64  
 6   days_since_prior_order  0 non-null      float64
 7   new_customer            0 non-null      bool   
dtypes: bool(1), float64(1), int64(5), object(1)
memory usage: 0.0+ bytes


#### as we can se no duplicates are found 

In [83]:
# 9. export final prods and ords data as scv into prepared Data with appropriate names 

df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'df_prods_cleaned_no_dups_26_12.csv'))

In [39]:
df_ords_clean.to_csv(os.path.join(path, '02 Data','Prepared Data', 'df_ords_cleaned_no_dups_26_12.csv'))